rfcs/45-stageless.md

```diff
@@ -0,0 +1,540 @@
+    .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+    // The `run_if_resource_is` method is convenient syntactic sugar that generates a run criteria
+    // for when you want to check the value of a resource (commonly an enum)
+    .add_system(gravity.run_if_resource_equals(Gravity::Enabled))
+    // Run criteria can be attached to labels: a copy of the run criteria will be applied to each system with that label
+    .configure_label(GameLabel::Physics.run_if_resource_equals(Paused(false)))
+    .run();
+}
+```
+
+**When multiple run criteria are attached to the same system, the system will run if and only if all of those run criteria return true.**
+
+Run criteria are evaluated "just before" the system that is attached to is run.
+The state that they read from will always be "valid" when the system that they control is being run: this is important to avoid strange bugs caused by race conditions and non-deterministic system ordering.
+It is impossible, for example, for a game to be paused *after* the run criteria checked if the game was paused but before the system that relied on the game not being paused completes.
+In order to understand exactly what this means, we need to understand atomic ordering constraints.
+
+### Atomic ordering constraints
+
+**Atomic ordering constraints** are a form of ordering constraints that ensure that two systems are run directly after each other.
+At least, as far as "directly after" is a meaningful concept in a parallel, nonlinear ordering.
+Like the atom, they are indivisible: nothing can get between them!
+
+To be fully precise, if we have system `A` that runs "atomically before" system `B` (`.add_system(a.atomically_before(b))`):
+
+- `A` is strictly before `B`
+- the data locks (in terms of read and write accesses to the `World`) of system `A` will not be released until system `B` has completed
+- the data locks from system `A` are ignored for the purposes of checking if system `B` can run
+  - after all, they're not actualy locked, merely reserved
+  - competing read-locks caused by other systems will still need to complete before a data store can be mutated; the effect is to "ignore the locks of `A`", not "skip the check completely"
+
+In addition to their use in run criteria, atomic ordering constraints are extremely useful for forcing tight interlocking behavior between systems.
+They allow us to ensure that the state read from and written to the earlier system cannot be invalidated.
+This functionality is extremely useful when updating indexes (used to look up components by value):
+allowing you to ensure that the index is still valid when the system using the index uses it.
+
+Of course, atomic ordering constraints should be used thoughtfully.
+As the strictest of the three types of system ordering dependency, they can easily result in unsatisfiable schedules if applied to large groups of systems at once.

```

Comment:
Why would atomic constraints lead to unsatisfiable schedules? Don't they just act like one system to external systems?